### PR TITLE
feat(runtime/worker): Structured cloning worker message passing.

### LINKED ISSUE #3557: Implement proper structured cloning between Workers
This is actually #3317, but we need a clearer issue name for tracking purposes.

Structured cloning implementation might be very useful in ways for transferring tasks between `Worker`s for load balancing, through low-cost transfer of `ArrayBuffer` (or even `SharedArrayBuffer`). For example, it might allow an HTTP request to move to another vacant thread if it somehow takes heavy CPU time to respond.

### PR DESCRIPTION
**Pull request by inteon**
Refactored webworkers to listen for host messages via an "async bin op" instead of directly injecting JSON into the runtime.

Fixes #3557.

### CODE REVIEW

**Comment by bartlomieju** [comment]
> Refactored webworkers to listen for host messages via an "async json op" instead of directly injecting JSON into the runtime.
> First steps towards #3557, main goal is to change the message passing, so instead of JSON SharedArrayBuffers can be used (not yet part of this pull request).
> [WIP] because a waker still has to be re-added to `poll_event_loop`
> ↳ need some more time to figure out how this works (feel free to help)
> 
> Would be nice if some intermediate feedback could be provided already, on both the approach and the current code.

Thanks for the PR, but the current design is on purpose. I'd rather not change code before there's a working prototype of structured cloned message passing and currently we're still missing appropriate bindings in `deno_core` to do that (which is in my backlog).

> NOTE: code also contains some unrelated fixes for linting 

It's not a good idea to throw such unrelated changes into a PR. Besides current formatting and linting passes so I'm not sure how that's better.

**Comment by inteon** [comment]
> > Refactored webworkers to listen for host messages via an "async json op" instead of directly injecting JSON into the runtime.
> > First steps towards #3557, main goal is to change the message passing, so instead of JSON SharedArrayBuffers can be used (not yet part of this pull request).
> > [WIP] because a waker still has to be re-added to `poll_event_loop`
> > ↳ need some more time to figure out how this works (feel free to help)
> > Would be nice if some intermediate feedback could be provided already, on both the approach and the current code.
> 
> Thanks for the PR, but the current design is on purpose. I'd rather not change code before there's a working prototype of structured cloned message passing and currently we're still missing appropriate bindings in `deno_core` to do that (which is in my backlog).
> 
> > NOTE: code also contains some unrelated fixes for linting
> 
> It's not a good idea to throw such unrelated changes into a PR. Besides current formatting and linting passes so I'm not sure how that's better.

Thank you for the feedback! I reverted the linting changes. Also, I will change the goal of this PR to create a structured cloning prototype. Like you said, that will make it easier to discuss what can be done better/ differently in my approach. While still making structured cloning work.
Then, the prototype can be used as a starting point to create the needed PRs to implement this feature (probably will involve multiple PRs).

Will update soon.

**Comment by inteon** [comment]
@bartlomieju I pushed a full prototype, please let me know what you think.

(BTW: I'll check tomorrow if any tests failed)

**Comment by inteon** [comment]
Current unsolved challenges in this prototype:
- `racy_worker` test is failing (not sure if something is wrong with test or with prototype)
- instead of JSON encoding arraybuffer, which was generated by the serialize instruction, create `transfer_to_worker_op`
- use same  `transfer_to_worker_op` to transfer other transferables using `post_message` #8341
- reevaluate use of  `op_worker_get_message`
- remove `core::mem::transmute` in "serialize_deserialize.rs"

File: core/serialize_deserialize.rs
Comment (bartlomieju) [inline core/serialize_deserialize.rs:91]: This bit needs to be addressed before landing; can you reproduce this problem as a standalone unit test?

File: core/serialize_deserialize.rs
Comment (bartlomieju) [inline core/serialize_deserialize.rs:71]: What needs to be fixed specifically?

File: core/serialize_deserialize.rs
Comment (bartlomieju) [inline core/serialize_deserialize.rs:1]: This file needs a copyright and some unit tests

File: core/bindings.rs
Comment (bartlomieju) [inline core/bindings.rs:55]: These two bindings need some tests, please add a new file similar to `core/encode_decode_test.js` that exercises these bindings

**Review by bartlomieju — COMMENTED** [review]
@inteon great work! There's a lot going on in this PR and it might be hard to land it in one go. Changes to `core/` look great, there are a few minor comments I have. 

I'm yet to properly go through changes to worker implementation; could you open a new PR with only changes to the `core/` directory so we can land them first?

File: core/serialize_deserialize.rs
Comment (inteon) [reply core/serialize_deserialize.rs:71]: Structured cloning allows normally for transfering arraybuffers, this still has to be implemented. It might be better to start with an implementation that does not yet support this.

File: core/serialize_deserialize.rs
Comment (inteon) [reply core/serialize_deserialize.rs:91]: The problem here is that the returned value from "value_deserializer.read_value(...)" should be able to outlive value_deserializer.

Comment (inteon) [reply core/serialize_deserialize.rs:91]: I created an issue for this: https://github.com/denoland/rusty_v8/issues/607.

File: core/bindings.rs
Comment (inteon) [reply core/bindings.rs:55]: I added some tests in this PR https://github.com/denoland/deno/pull/9458.

File: core/serialize_deserialize.rs
Comment (bartlomieju) [reply core/serialize_deserialize.rs:71]: Yes, that makes sense, let's start with providing only structured clone and worry about transfers later

**Comment by inteon** [comment]
Waiting for #9458 to land, also improvements to ops_messages are desired (see #9433 and #9457).

**Comment by bartlomieju** [comment]
@inteon let me know if this PR is ready, I don't think we'll land it for 1.8 though

**Comment by inteon** [comment]
> @inteon let me know if this PR is ready, I don't think we'll land it for 1.8 though

Indeed, let's not rush things. First lets make sure it works well.

**Comment by inteon** [comment]
Blocked on #9457

File: runtime/js/01_errors.js
Comment (bartlomieju) [inline runtime/js/01_errors.js:144]: I'm not happy with this error - all errors defined in this file are available in `Deno.errors` but this one is not. Additionally I'm pretty sure that workers should throw plain `Error` or maybe even `DOMException`. @nayeemrmn thoughts?

File: runtime/js/99_main.js
Comment (bartlomieju) [inline runtime/js/99_main.js:82]: What's the advantage of polling for messages from within runtime instead of explicitly calling a function like `workerMessageRecvCallback`?

Also, please rename to `pollForMessages`.

File: runtime/ops/web_worker.rs
Comment (bartlomieju) [inline runtime/ops/web_worker.rs:25]: Maybe the receiver could be part of `WebWorkerHandle`?

File: runtime/ops/web_worker.rs
Comment (bartlomieju) [inline runtime/ops/web_worker.rs:78]: It should be just `lock`; if there are multiple `op_worker_get_message` ops dispatched it should be a panic indicating problem with internal code.

File: runtime/ops/web_worker.rs
Comment (bartlomieju) [inline runtime/ops/web_worker.rs:100]: `NotConnected` is an IO error and we shouldn't use it for this purpose; it should be some internal type instead

File: runtime/ops/worker_host.rs
Comment (bartlomieju) [inline runtime/ops/worker_host.rs:494]: Ditto

File: runtime/ops/web_worker.rs
Comment (inteon) [reply runtime/ops/web_worker.rs:78]: Here this is used too, is this different or should it change too?
https://github.com/denoland/deno/blob/26f7a3f185881c12081eb6d3cedbf9e637376d86/runtime/web_worker.rs#L76-L77

File: runtime/ops/web_worker.rs
Comment (inteon) [reply runtime/ops/web_worker.rs:25]: Receiver is now part of "WorkerChannelsInternal" its the 'other end' of the communication line. WebWorkerHandle is only used to terminate the worker from within the worker.

Comment (inteon) [reply runtime/ops/web_worker.rs:25]: Normally 'WebWorkerHandle' should be only used by the host in my opinion.

File: runtime/ops/web_worker.rs
Comment (inteon) [reply runtime/ops/web_worker.rs:100]: Can I register a new error class in 01_errors.js for this, or should I create an error class in 11_worker.js. Not sure what best practices are here.

File: runtime/js/99_main.js
Comment (inteon) [reply runtime/js/99_main.js]: I renamed the function to `pollForMessages`.
I think its simpler to just register an ops function, also this is how the host receives messages from the workers too.

File: runtime/ops/web_worker.rs
Comment (inteon) [reply runtime/ops/web_worker.rs]: I created a WorkerConnectionError for this, also this error triggers if --allow-read is not provided in combination with workers.

**Comment by bartlomieju** [comment]
@inteon please rebase when you get a moment, let's land this feature for 1.10

**Comment by inteon** [comment]
@bartlomieju finished the rebase, PR also includes some small worker related cleanup (changed wrong text in error & moved bench). I also revised WebWorkerHandle and WebWorkerInternalHandle to give better separation between cross-tread and same-thread handle.
NOTE: the ops-signatures probably are not yet optimal, I propose we optimize them in a follow-up PR

File: runtime/ops/web_worker.rs
Comment (bnoordhuis) [inline runtime/ops/web_worker.rs:41]:
```suggestion
          Ok(maybe_data.unwrap_or_default())
```

File: runtime/web_worker.rs
Comment (bnoordhuis) [inline runtime/web_worker.rs:40]: Consider turning this into a newtype (`pub struct WorkerId(u32);`) if you're worried about people mixing up ids with other integral values.

File: runtime/web_worker.rs
Comment (bnoordhuis) [inline runtime/web_worker.rs:65]:
```suggestion
    // the worker must have terminated but the termination message has not yet been received.
```
(Also: long line. I guess I can't complain if the linter doesn't but consider wrapping it at 80 columns.)

File: runtime/web_worker.rs
Comment (bnoordhuis) [inline runtime/web_worker.rs:589]: Consider using serde_v8 to encode the value, here and below. The wire format isn't stable.

**Review by bnoordhuis — APPROVED** [review]

File: runtime/web_worker.rs
Comment (bnoordhuis) [reply runtime/web_worker.rs:589]: Sorry, yes. I suppose that makes it inconvenient to call out to. Consider it a very light suggestion then.

File: runtime/web_worker.rs
Comment (inteon) [reply runtime/web_worker.rs]: Fixed it

File: runtime/web_worker.rs
Comment (inteon) [reply runtime/web_worker.rs:607]: I added some TODO comments here, if accessing `Deno.core.serialize()` from rust would become easier this can be fixed; also depends on how often it breaks due to v8 internals changing

File: runtime/js/11_workers.js
Comment (bartlomieju) [inline runtime/js/11_workers.js:243]: Nitpick: this op is called from the "host" side, so I'd prefer to keep `op_host_` prefix

File: runtime/js/11_workers.js
Comment (bartlomieju) [inline runtime/js/11_workers.js:263]: Ditto

File: runtime/js/99_main.js
Comment (bartlomieju) [inline runtime/js/99_main.js:86]: Isn't call to `new Uint8Array` superfluous here?

File: runtime/ops/web_worker.rs
Comment (bartlomieju) [inline runtime/ops/web_worker.rs:19]: @inteon could you please rewrite those op handlers to be functions in this module instead of closures? It's becoming hard to read them

File: runtime/ops/worker_host.rs
Comment (bartlomieju) [inline runtime/ops/worker_host.rs:486]: Please revert - otherwise there will only be a number in the thread name

File: runtime/web_worker.rs
Comment (bartlomieju) [inline runtime/web_worker.rs:324]: Ditto, please revert

File: runtime/web_worker.rs
Comment (bartlomieju) [inline runtime/web_worker.rs:428]: Great to see all of that code go!

**Review by bartlomieju — COMMENTED** [review]
@inteon thank you so much for this PR, it's really great that you managed to implement all of that functionality in so little code. I have only some minor/nitpicky comments, other than that it's ready to land. Great to see this landed for 1.10!

File: runtime/js/99_main.js
Comment (lucacasonato) [reply runtime/js/99_main.js:86]: I thought `Box<[u8]>` is auto serialized to Uint8Array... cc @AaronO

Comment (lucacasonato) [reply runtime/js/99_main.js:86]: Actually just checked - `ZeroCopyBuf` / `serde_v8::MagicBuffer` is auto serialized to a `Uint8Array`. You can `ZeroCopyBuf::from(Box<[u8]>)`.

File: runtime/ops/worker_host.rs
Comment (inteon) [reply runtime/ops/worker_host.rs:486]: @bartlomieju please let me know if this is ok, or if I should still change it

File: runtime/js/99_main.js
Comment (inteon) [reply runtime/js/99_main.js]: I fixed this in my latest changes, also at the host side this was fixed.

**Comment by inteon** [comment]
I optimized the worker <-> host message passing ops (removed all places where buffers were copied).
@AaronO I also made some small changes to serde_v8, please check if these are acceptable

File: runtime/ops/worker_host.rs
Comment (bartlomieju) [reply runtime/ops/worker_host.rs:486]: I see, that sounds good

**Review by bartlomieju — APPROVED** [review]
LGTM as well ✅  thanks for seeing this through @inteon, this is a massive feature that was highly requested.

File: runtime/js/11_workers.js
Comment (ry) [inline runtime/js/11_workers.js:61]: \o/

File: runtime/js/99_main.js
Comment (ry) [inline runtime/js/99_main.js:85]: I think it'd be more readable to write
```js
const bufferMsg = await core.opAsync("op_worker_get_message");
```

File: runtime/js/99_main.js
Comment (ry) [inline runtime/js/99_main.js:79]: Maybe just
```js
core.opSync("op_worker_post_message", null, dataIntArray);
```

File: serde_v8/src/magic/buffer.rs
Comment (ry) [inline serde_v8/src/magic/buffer.rs:17]: If a Mutex instead of a Cell is used, MagicBuffer will automatically inherit Sync.
```diff
diff --git a/serde_v8/src/magic/buffer.rs b/serde_v8/src/magic/buffer.rs
index c14284c7..7ffff861 100644
--- a/serde_v8/src/magic/buffer.rs
+++ b/serde_v8/src/magic/buffer.rs
@@ -1,9 +1,9 @@
 use rusty_v8 as v8;
 
-use std::cell::Cell;
 use std::fmt;
 use std::ops::Deref;
 use std::ops::DerefMut;
+use std::sync::Mutex;
 
 use super::zero_copy_buf::ZeroCopyBuf;
 
@@ -11,11 +11,9 @@ use super::zero_copy_buf::ZeroCopyBuf;
 // allowing us to use a single type for familiarity
 pub enum MagicBuffer {
   FromV8(ZeroCopyBuf),
-  ToV8(Cell<Option<Box<[u8]>>>),
+  ToV8(Mutex<Option<Box<[u8]>>>),
 }
 
-unsafe impl Sync for MagicBuffer {}
-
 impl MagicBuffer {
   pub fn new<'s>(
     scope: &mut v8::HandleScope<'s>,
@@ -36,7 +34,7 @@ impl Clone for MagicBuffer {
 
 impl Default for MagicBuffer {
   fn default() -> Self {
-    MagicBuffer::ToV8(Cell::new(Some(vec![0_u8; 0].into_boxed_slice())))
+    MagicBuffer::ToV8(Mutex::new(Some(vec![0_u8; 0].into_boxed_slice())))
   }
 }
 
@@ -73,7 +71,7 @@ impl DerefMut for MagicBuffer {
 
 impl From<Box<[u8]>> for MagicBuffer {
   fn from(buf: Box<[u8]>) -> Self {
-    MagicBuffer::ToV8(Cell::new(Some(buf)))
+    MagicBuffer::ToV8(Mutex::new(Some(buf)))
   }
 }
 
@@ -100,7 +98,7 @@ impl serde::Serialize for MagicBuffer {
         let value: &[u8] = &*buf;
         value.into()
       }
-      Self::ToV8(x) => x.take().expect("MagicBuffer was empty"),
+      Self::ToV8(x) => x.lock().unwrap().take().expect("MagicBuffer was empty"),
     };
     let hack: [usize; 2] = unsafe { std::mem::transmute(boxed) };
     let f1: u64 = hack[0] as u64;

```

File: serde_v8/src/magic/buffer.rs
Comment (ry) [inline serde_v8/src/magic/buffer.rs:41]: I don't agree that the Default MagicBuffer should be an empty buffer like this.

I think it makes more sense if an empty buffer is explicitly constructed in op_worker_get_message() (runtime/ops/web_worker.rs:57). That is the only place `MagicBuffer::default()` is used.

File: serde_v8/src/magic/buffer.rs
Comment (piscisaureus) [inline serde_v8/src/magic/buffer.rs:100]: The buffer gets copied here. This seems less optimal - is it intentional?

File: runtime/web_worker.rs
Comment (piscisaureus) [inline runtime/web_worker.rs:462]: "terminated"

File: serde_v8/src/magic/buffer.rs
Comment (piscisaureus) [inline serde_v8/src/magic/buffer.rs:17]: You can't implement `Sync` on a `Cell<Option<Box<[u8]>>>`.

Cell is probably not dangerous in practice here, but misleading because a Cell is very specifically not Sync. There's `UnsafeCell` for that.

Mutable `Option<Box<T>>` is definitely not ok to make `Sync`.

Comment (bartlomieju) [reply serde_v8/src/magic/buffer.rs:17]: Applied

File: serde_v8/src/magic/buffer.rs
Comment (inteon) [reply serde_v8/src/magic/buffer.rs:96]: No, this is not intentional; lets see if I can fix this.

Comment (bartlomieju) [reply serde_v8/src/magic/buffer.rs:96]: @inteon I already fixed it, used `&buf` instead

File: runtime/js/99_main.js
Comment (leopiccionia) [reply runtime/js/99_main.js]: An awaited sync value is converted to a resolved Promise (see [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await)). It typically means that the function execution is suspended and a microtask is enqueued, but I don't know the internals of Deno's event loop.

A common idiom is:
```js
let result = task()
if (isPromise(result)) {
    result = await result
}
```

**Pull request merged**
